### ISSUE TITLE
react hooks

### ISSUE THREAD

**Issue by tx4x** [OneArmyWorld/onearmy#252]
could we please agree on React@next regarding hooks? After SFC, they added['hooks'](https://www.youtube.com/watch?v=wXLf18DsV-I) which is soo much cleaner and easier to maintain than all this componentUnmout/unsubscribe bloat :-)

the concept/pattern of classes/oop in the JS world is nowadays obsolete (happy to see that after all). implementing a class comes also with some other quirks, not really needed. 

at some point i'd love to see this code base also making use of Ramda or whatever suits you. 

thanks, and yes, i know i can use whatever i like and propose but there are also a few changes i'd like to do do to other components, and whilst at it, get rid of old habbits :-)

**Comment by tx4x** [comment]
if you're not familiar, please check before 'functional programming' videos on youtube, especially the part of getting rid of effects, abstractions and composition

**Comment by chrismclarke** [comment]
https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both

I think as per react's own docs... hooks are still not fully battle-tested (particularly for typescript). I think eventually most components will move to use them (I'm a big fan, and I think there's a general trend towards them so makes sense to be somewhat opinionated) but I think it makes sense to first create a couple good examples for the newer components to help people see how we use them.

As for OO vs functional and anything in between - I think this is much more personal taste so again feel free to make use of what makes you productive and if it encourages more people to do the same then that's an obvious bonus.

Also if you have any specific video recommendations would always appreciate

**Comment by tx4x** [comment]
hi,
thanks! Ok then, I will go ahead for now with react-hooks, there is still plenty to do : https://github.com/tx4x/onearmy/blob/module--discussions/src/components/Editor/plugins/image/ui/Dialog.tsx

**Comment by BenGamma** [comment]
Hooks are finally released as stable, but there is some breaking changes from alphas versions 👉 https://reactjs.org/blog/2019/02/06/react-v16.8.0.html#hooks-changelog-since-alpha-versions

**Comment by chrismclarke** [comment]
Good to know, suggest we give it a couple weeks before upgrading for people to iron out the kinks

**Comment by tx4x** [comment]
@chrismclarke , @BenGamma 
hi,
sorry this was an exceptional busy week for me. I could manage to dedicate 6 hours this week and 2 full days next week. I expect the most of the editor to be done by then - in a way that you can integrate this where needed. I will have also to make another react@next PR to get hooks@stable.
see you monday
g

**Comment by chrismclarke** [comment]
v18.0.1 came through so I made a PR to update. Quite quickly I realised things actually aren't quite so simple as major updates also contains a lot of changes to react app scripts. It would be fine if we were still using them, however the platform has been ejected and so requires manual update.

Am working on this now, should hopefully have up by the end of the day

**Comment by tx4x** [comment]
yep, i noticed this too in my trials to get webpack to latest which i gave up after 4+ hours of banging around. there were quite some webpack plugins which had real issues

what was the reason to 'eject' anyways, may be you go back to defaults ?

**Comment by chrismclarke** [comment]
react upgrade now implemented, feel free to use as desired

**Comment by chrismclarke** [comment]
I think I spent about 2 hours headbanging but got there in the end! We initially ejected to support SCSS loading, however that is now both deprecated as we used styled-components and because the latest scripts includes a loader. 

I saw custom there is still custom scripts for enzyme testing within the ejected config which was the main reason I wanted to keep ejected. This might be reviewed as we move forward with testing, but would need to speak to @PepijnSenders to understand if we want to keep using enzyme and whether it could work with a standard create-react-app set of scripts.
